### ISSUE TITLE
Scheduled job to delete non activated users fails 

### ISSUE THREAD

**Issue by nivemaham** [RADAR-base/ManagementPortal#280]
Scheduled job to delete non activated users fails with following exception. 
The job should check whether it is a participant or other users and delete entries in corresponding tables to avoid foreign-key violations 
**Enviornment**
test
**Date reported**
03-07-2018
**Stack trace**
```
managementportal-app_1         | 2018-07-03 01:00:00.113 ERROR 8 --- [pool-5-thread-1] o.s.s.s.TaskUtils$LoggingErrorHandler    : Unexpected error occurred in scheduled task.
managementportal-app_1         | 
managementportal-app_1         | org.springframework.dao.DataIntegrityViolationException: could not execute statement; SQL [n/a]; constraint [fk_subject_user_id]; nested exception is org.hibernate.exception.ConstraintViolationException: could not execute statement
managementportal-app_1         | 	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:278)
managementportal-app_1         | 	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:244)
managementportal-app_1         | 	at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:521)
managementportal-app_1         | 	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:761)
managementportal-app_1         | 	at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:730)
managementportal-app_1         | 	at org.springframework.transaction.interceptor.TransactionAspectSupport.commitTransactionAfterReturning(TransactionAspectSupport.java:518)
managementportal-app_1         | 	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:292)
managementportal-app_1         | 	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
managementportal-app_1         | 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
managementportal-app_1         | 	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:673)
managementportal-app_1         | 	at org.radarcns.management.service.UserService$$EnhancerBySpringCGLIB$$1572bbe9.removeNotActivatedUsers(<generated>)
managementportal-app_1         | 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
managementportal-app_1         | 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
managementportal-app_1         | 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
managementportal-app_1         | 	at java.lang.reflect.Method.invoke(Method.java:498)
managementportal-app_1         | 	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:65)
managementportal-app_1         | 	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
managementportal-app_1         | 	at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:81)
managementportal-app_1         | 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
managementportal-app_1         | 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
managementportal-app_1         | 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
managementportal-app_1         | 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
managementportal-app_1         | 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
managementportal-app_1         | 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
managementportal-app_1         | 	at java.lang.Thread.run(Thread.java:748)
managementportal-app_1         | Caused by: org.hibernate.exception.ConstraintViolationException: could not execute statement
managementportal-app_1         | 	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:112)
managementportal-app_1         | 	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:42)
managementportal-app_1         | 	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:111)
managementportal-app_1         | 	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:97)
managementportal-app_1         | 	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:178)
managementportal-app_1         | 	at org.hibernate.engine.jdbc.batch.internal.NonBatchingBatch.addToBatch(NonBatchingBatch.java:45)
managementportal-app_1         | 	at org.hibernate.persister.entity.AbstractEntityPersister.delete(AbstractEntityPersister.java:3325)
managementportal-app_1         | 	at org.hibernate.persister.entity.AbstractEntityPersister.delete(AbstractEntityPersister.java:3562)
managementportal-app_1         | 	at org.hibernate.action.internal.EntityDeleteAction.execute(EntityDeleteAction.java:99)
managementportal-app_1         | 	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:589)
managementportal-app_1         | 	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:463)
managementportal-app_1         | 	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:337)
managementportal-app_1         | 	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:39)
managementportal-app_1         | 	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1437)
managementportal-app_1         | 	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:493)
managementportal-app_1         | 	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3207)
managementportal-app_1         | 	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2413)
managementportal-app_1         | 	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:473)
managementportal-app_1         | 	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.beforeCompletionCallback(JdbcResourceLocalTransactionCoordinatorImpl.java:156)
managementportal-app_1         | 	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.access$100(JdbcResourceLocalTransactionCoordinatorImpl.java:38)
managementportal-app_1         | 	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commit(JdbcResourceLocalTransactionCoordinatorImpl.java:231)
managementportal-app_1         | 	at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:68)
managementportal-app_1         | 	at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:517)
managementportal-app_1         | 	... 22 common frames omitted
managementportal-app_1         | Caused by: org.postgresql.util.PSQLException: ERROR: update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
managementportal-app_1         |   Detail: Key (id)=(14257) is still referenced from table "subject".
managementportal-app_1         | 	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2455)
managementportal-app_1         | 	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2155)
managementportal-app_1         | 	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:288)
managementportal-app_1         | 	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:430)
managementportal-app_1         | 	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:356)
managementportal-app_1         | 	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:168)
managementportal-app_1         | 	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:135)
managementportal-app_1         | 	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
managementportal-app_1         | 	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
managementportal-app_1         | 	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:175)
managementportal-app_1         | 	... 40 common frames omitted

```

**Comment by blootsvoets** [comment]
Perhaps cascading should be implemented for the foreign key? i.e. `ON DELETE CASCADE` for `radar_user.id`. See https://www.postgresql.org/docs/9.6/static/ddl-constraints.html.

**Comment by yatharthranjan** [comment]
Yes on the prod i deleted some test subjects and got these logs in postgres -
```
ERROR:  update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
DETAIL:  Key (id)=(10776) is still referenced from table "subject".
STATEMENT:  delete from radar_user where id=$1
ERROR:  update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
DETAIL:  Key (id)=(10776) is still referenced from table "subject".
STATEMENT:  delete from radar_user where id=$1
ERROR:  update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
DETAIL:  Key (id)=(1858) is still referenced from table "subject".
STATEMENT:  delete from radar_user where id=$1
ERROR:  update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
DETAIL:  Key (id)=(1858) is still referenced from table "subject".
STATEMENT:  delete from radar_user where id=$1
ERROR:  update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
DETAIL:  Key (id)=(1858) is still referenced from table "subject".
STATEMENT:  delete from radar_user where id=$1
ERROR:  update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
DETAIL:  Key (id)=(1858) is still referenced from table "subject".
STATEMENT:  delete from radar_user where id=$1
ERROR:  update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
DETAIL:  Key (id)=(1858) is still referenced from table "subject".
STATEMENT:  delete from radar_user where id=$1
ERROR:  update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
DETAIL:  Key (id)=(1858) is still referenced from table "subject".
STATEMENT:  delete from radar_user where id=$1
ERROR:  update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
DETAIL:  Key (id)=(1858) is still referenced from table "subject".
STATEMENT:  delete from radar_user where id=$1
ERROR:  update or delete on table "radar_role" violates foreign key constraint "fk_role_users_roles_id" on table "role_users"
DETAIL:  Key (id)=(2752) is still referenced from table "role_users".
STATEMENT:  delete from radar_role where id=$1
ERROR:  update or delete on table "radar_role" violates foreign key constraint "fk_role_users_roles_id" on table "role_users"
DETAIL:  Key (id)=(2752) is still referenced from table "role_users".
STATEMENT:  delete from radar_role where id=$1
ERROR:  update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
DETAIL:  Key (id)=(1858) is still referenced from table "subject".
STATEMENT:  delete from radar_user where id=$1
ERROR:  update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
DETAIL:  Key (id)=(1858) is still referenced from table "subject".
STATEMENT:  delete from radar_user where id=$1
ERROR:  update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
DETAIL:  Key (id)=(1858) is still referenced from table "subject".
STATEMENT:  delete from radar_user where id=$1
ERROR:  update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
DETAIL:  Key (id)=(1858) is still referenced from table "subject".
STATEMENT:  delete from radar_user where id=$1
ERROR:  update or delete on table "radar_user" violates foreign key constraint "fk_subject_user_id" on table "subject"
DETAIL:  Key (id)=(1858) is still referenced from table "subject".
STATEMENT:  delete from radar_user where id=$1
```

**Comment by dennyverbeeck** [comment]
Fixed by #323
